### PR TITLE
Fix poisson_nll_loss with full option

### LINKED ISSUE #28575: PoissonNLLLoss does not compute the value documented
## 🐛 Bug
When the forward pass of torch.nn.PoissonNLLLoss does not compute the same value as described in the documentation.

## To Reproduce

```
import math
import torch


inp = 3.
target = 5.

nll_poisson_math = inp - target * math.log(inp) + math.log(math.factorial(target))
nll_poisson_stirling_math = inp - target * math.log(inp) + target * math.log(target) - target + 0.5 * math.log(2 * math.pi * target)

nll_poisson_loss = torch.nn.PoissonNLLLoss(log_input=False, 
                        full=True,
                        size_average=None,
                        eps=0.,
                        reduce=None,
                        reduction='none')

inpTensor, targetTensor = torch.as_tensor([inp]), torch.as_tensor([target])
nll_poisson_pytorch = nll_poisson_loss(log_input=inpTensor, target=targetTensor)

print(nll_poisson_math, nll_poisson_stirling_math, nll_poisson_pytorch.item())
```

2.2944302994414967 2.2777856082516754 -2.4930615425109863
## Environment

pytorch 1.3.0

### PR DESCRIPTION
**Pull request by hi-ogawa**
This fixes https://github.com/pytorch/pytorch/issues/28575.

It seems `poisson_nll_loss` was implemented with the incorrect assumption about `masked_select`, which actually doesn't return tensor with the same storage, so in-place operation used there didn't work as intended.
Here I used `masked_fill` instead.

Also, the existing test didn't have `reference_fn`, so I added it (although it's not fundamentally useful since current cpp `poisson_nll_loss` itself does exactly same algorithm as `reference_fn`).

Thanks in advance for reviewing this PR.

### CODE REVIEW

File: test/common_nn.py
Comment (hi-ogawa) [inline test/common_nn.py]: I found that `criterion_tests` doesn't test backward by default. So, I added `check_forward_only=False` explicitly here for easier code review.
But, I suppose, such default is intended for faster test execution time, so I'll remove this if it's desired so.

Comment (albanD) [reply test/common_nn.py]: @mruberry do you know what is our take on this? Should we set this back to the default before merging?

**Review by albanD — APPROVED** [review]
This looks good, thanks for the PR !
Let's wait for @mruberry's answer for the tests and then we can merge.

**Comment by albanD** [comment]
@hi-ogawa  The python linter complains that you should put 2 spaces before the inline comments in the test file. Could you add that quickly please?

**Comment by hi-ogawa** [comment]
@albanD thanks a lot for the comment. I just fixed lint failures and flake CI looks okay now.

**Comment by albanD** [comment]
Looks good. Thanks for the quick update.

File: test/common_nn.py
Comment (mruberry) [reply test/common_nn.py]: It looks like PyTorch tests backwards in module_tests and not criterion_tests (per @hi-ogawa's comment). Without a compelling reason I wouldn't change this behavior. NLLLoss is also tested by the module tests. (Unfortunately our test suite is a confusing mess of redundancy.)

Comment (albanD) [reply test/common_nn.py]: I guess the point here is that since the function itself is autodiffed, there is no point to check the gradient computations.
@hi-ogawa could you set this flag back to the default value (remove the line).

Comment (hi-ogawa) [reply test/common_nn.py]: @mruberry @albanD Thanks a lot for the review. Right, `poissonnllloss_no_reduce_test` (Line 280) is included in `new_module_tests` instead of `new_criterion_tests` and that one actually checks backward. Okay, I removed `check_forward_only=False` to align with other tests here and I hope it's ready to merge now.

**Comment by albanD** [comment]
Thanks for updating.
Merging now.

**Pull request closed without merging**
